tests/jax/core_system_components/parameter_client_test.py

```diff
@@ -15,4 +15,406 @@
+        "key_0": np.array(0, dtype=np.int32),
+        "key_1": np.array(2, dtype=np.float32),
+        "key_2": np.array(2, dtype=np.int32),
+        "key_3": np.array(4, dtype=np.int32),
+        "key_4": np.array(5, dtype=np.int32),
+        "networks-network_key_0": {"layer_0": {"weights": 1, "biases": 1}},
+        "networks-network_key_1": {"layer_0": {"weights": 2, "biases": 2}},
+        "networks-network_key_2": {"layer_0": {"weights": 3, "biases": 3}},
+        "allkey_0": np.array(0, dtype=np.int32),
+    }
+    assert parameter_client._get_call_counter == 0
+    assert parameter_client._get_future is None
+
+
+def test_set_async(parameter_client: ParameterClient) -> None:
+    """Set set async method"""
+
+    # test that set_call_counter is incremented
+    parameter_client.set_async()
+    assert parameter_client._set_call_counter == 1
+
+    # set call counter equal to update period and verify the request was made
+    parameter_client._set_call_counter = 10
+
+    # increment set parameters, call set_async and verify that
+    # parameters were set asynchronously
+    increment_set_parameters(
+        params=parameter_client._parameters, names=parameter_client._set_keys
+    )
+
+    parameter_client.set_async()
+
+    assert parameter_client._client.store._set_params == {
+        "key_0": np.array(1, dtype=np.int32),
+        "key_2": np.array(3, dtype=np.int32),
+    }
+    assert parameter_client._set_call_counter == 0
+
+    # assert that _set_future is None again after request was made
+    parameter_client.set_async()
+    assert parameter_client._set_future is None
+
+
+def test_set_and_get_async(parameter_client: ParameterClient) -> None:
+    """Set set async method"""
+
+    # test that set_and_get_call_counter is incremented
+    parameter_client.set_and_get_async()
+    assert parameter_client._set_get_call_counter == 1
+
+    # set counter equal to update period and verify the request was made
+    parameter_client._set_get_call_counter = 10
+
+    # increment set parameters
+    increment_set_parameters(
+        params=parameter_client._parameters, names=parameter_client._set_keys
+    )
+
+    parameter_client.set_and_get_async()
+
+    assert parameter_client._parameters == {
+        "key_0": np.array(1, dtype=np.int32),
+        "key_1": np.array(2, dtype=np.float32),
+        "key_2": np.array(3, dtype=np.int32),
+        "key_3": np.array(4, dtype=np.int32),
+        "key_4": np.array(5, dtype=np.int32),
+        "networks-network_key_0": {"layer_0": {"weights": 1, "biases": 1}},
+        "networks-network_key_1": {"layer_0": {"weights": 2, "biases": 2}},
+        "networks-network_key_2": {"layer_0": {"weights": 3, "biases": 3}},
+        "allkey_0": np.array(0, dtype=np.int32),
+    }
+
+    assert parameter_client._set_get_call_counter == 0
+
+    parameter_client.set_and_get_async()
+
+    assert parameter_client._set_get_future is None
+
+
+def test_add_async(parameter_client: ParameterClient) -> None:
```

Comment:
I can't find a test here for what happens when `self._add_future is None`

Reply:
By default `self._add_future` is set to `None`, please see [here](https://github.com/instadeepai/Mava/blob/9a0835a4998d17cab580a7269443c12cc16cf6d1/mava/systems/jax/parameter_client.py#L87). So we technically check for that condition in lines 319 - 322. I have also added an extra assert to verify that `self._add_future.done()` is `True` after adding the parameters. I can also explicitly set the value of `self.add_future` to `None` in the test if you think that would make it more explicit?

Reply:
Ah I see, no worries then!
Maybe just add an assert before calling `add_async` that `self._add_future is None`, to ensure that the default is set correctly?

Reply:
Added it :call_me_hand: